Update shell lecture notes + add advanced shell (#5)
afnom/missing-semester: _2024/shell-intermediate.md

Both of the above can also be applied recursively to change the permissions of an entire directory structure. Do this with caution! For details on how to use these commands, take a look at their manual pages.

# Package managers

> The primary method of installing packages for linux is usually using a package manager from the command line. There is some variation of package manager based on linux distribution, but most use one of the following: `apt` (Debian-based), `pacman` (Arch-based), `dnf` (RedHat-based), or `zypper` (SUSE-based). For simplicity's sake, we'll be demonstrating how to use `apt`, as it is the most popular package manager, and the principles should carry on to the other package managers through different command names.

Reviewer: ```suggestion
The primary method of installing applications on Linux systems is by using a package manager. Each Linux distribution has its own package manager, but most use one of the following: `apt` (Debian-based), `pacman` (Arch-based), `dnf` (RedHat-based), or `zypper` (SUSE-based). For simplicity's sake, we'll be demonstrating how to use `apt`, as it is the most popular package manager, and the principles should carry on to the other package managers through different command names.
```

Since package managers edit package files system-wide, root privileges are needed to install, delete, and update package repositories, so most commands will be prefixed by `sudo`.

**Note: Take the same care with using `sudo` now as you always should.**

As as example, let's install `cowsay`:

First, try running `cowsay "Hello!"`. Usually, the `cowsay` command isn't installed, so `bash` should return a `comand not found` error.

> Let's start by updating our package repositories with `sudo apt update`. Run `sudo apt install cowsay` and enter your password to install the package. Folloing this, run `cowsay "Hello!"` and notice that the command is now available. To remove `cowsay`, run `sudo apt remove cowsay`.

Reviewer: Maybe it's better for paragraphs involving a lot of separate commands to be broken up using code blocks with output examples instead? E.g.

```bash
sudo apt update
# [...]

sudo apt install cowsay
# [sudo] password for user:

cowsay "Hello"
#  _______
# < Hello >
# -------
#        \   ^__^
#         \  (oo)\_______
#            (__)\       )\/\
#                ||----w |
#                ||     ||
```
